File: docs/static/board.js
// Core Sudoku board representation, validation, and puzzle generation.
// Port of sudoku_ai/board.py to run entirely client-side (GitHub Pages
// only serves static files, so all solving happens in the browser).
const Board = (() => {
  const SIZE = 9;
  const BOX = 3;

  function emptyGrid() {
    return Array.from({ length: SIZE }, () => Array(SIZE).fill(0));
  }

  function deepCopy(grid) {
    return grid.map((row) => row.slice());
  }

  function boxOrigin(r, c) {
    return [Math.floor(r / BOX) * BOX, Math.floor(c / BOX) * BOX];
  }

  function candidates(grid, r, c) {
    if (grid[r][c] !== 0) return new Set();
    const used = new Set(grid[r]);
    for (let i = 0; i < SIZE; i++) used.add(grid[i][c]);
    const [br, bc] = boxOrigin(r, c);
    for (let i = br; i < br + BOX; i++)
      for (let j = bc; j < bc + BOX; j++) used.add(grid[i][j]);
    const out = new Set();
    for (let v = 1; v <= 9; v++) if (!used.has(v)) out.add(v);
    return out;
  }

  function isValidPlacement(grid, r, c, val) {
    if (val === 0) return true;
    for (let i = 0; i < SIZE; i++) {
      if (i !== c && grid[r][i] === val) return false;
      if (i !== r && grid[i][c] === val) return false;
    }
    const [br, bc] = boxOrigin(r, c);
    for (let i = br; i < br + BOX; i++)
      for (let j = bc; j < bc + BOX; j++)
        if ((i !== r || j !== c) && grid[i][j] === val) return false;
    return true;
  }

  function isComplete(grid) {
    for (let r = 0; r < SIZE; r++)
      for (let c = 0; c < SIZE; c++) if (grid[r][c] === 0) return false;
    return true;
  }

  function conflictsCount(grid) {
    let count = 0;
    for (let i = 0; i < SIZE; i++) {
      const row = grid[i].filter((v) => v !== 0);
      count += row.length - new Set(row).size;
      const col = [];
      for (let r = 0; r < SIZE; r++) if (grid[r][i] !== 0) col.push(grid[r][i]);
      count += col.length - new Set(col).size;
    }
    for (let br = 0; br < SIZE; br += BOX) {
      for (let bc = 0; bc < SIZE; bc += BOX) {
        const vals = [];
        for (let i = br; i < br + BOX; i++)
          for (let j = bc; j < bc + BOX; j++) if (grid[i][j] !== 0) vals.push(grid[i][j]);
        count += vals.length - new Set(vals).size;
      }
    }
    return count;
  }

  function solveBacktrackingInPlace(grid) {
    for (let r = 0; r < SIZE; r++) {
      for (let c = 0; c < SIZE; c++) {
        if (grid[r][c] === 0) {
          const cands = Array.from(candidates(grid, r, c));
          for (const val of cands) {
            grid[r][c] = val;
            if (solveBacktrackingInPlace(grid)) return true;
            grid[r][c] = 0;
          }
          return false;
        }
      }
    }
    return true;
  }

  function shuffle(arr) {
    for (let i = arr.length - 1; i > 0; i--) {
      const j = Math.floor(Math.random() * (i + 1));
      [arr[i], arr[j]] = [arr[j], arr[i]];
    }
    return arr;
  }

  function generateFullSolution() {
    const grid = emptyGrid();
    for (let box = 0; box < SIZE; box += BOX) {
      const nums = shuffle([1, 2, 3, 4, 5, 6, 7, 8, 9]);
      let idx = 0;
      for (let i = 0; i < BOX; i++)
        for (let j = 0; j < BOX; j++) grid[box + i][box + j] = nums[idx++];
    }
    solveBacktrackingInPlace(grid);
    return grid;
  }

  function generatePuzzle(numClues = 32) {
    const solution = generateFullSolution();
    const puzzle = deepCopy(solution);
    const cells = [];
    for (let r = 0; r < SIZE; r++) for (let c = 0; c < SIZE; c++) cells.push([r, c]);
    shuffle(cells);
    const toRemove = SIZE * SIZE - numClues;
    for (let i = 0; i < toRemove; i++) {
      const [r, c] = cells[i];
      puzzle[r][c] = 0;
    }
    return { puzzle, solution };
  }

  function parseGrid(text) {
    const digits = text.split("").filter((ch) => /[0-9.]/.test(ch));
    if (digits.length !== 81) throw new Error(`Expected 81 cells, got ${digits.length}`);
    const grid = emptyGrid();
    digits.forEach((ch, idx) => {
      const r = Math.floor(idx / SIZE), c = idx % SIZE;
      grid[r][c] = ch === "." ? 0 : parseInt(ch, 10);
    });
    return grid;
  }

  return {
    SIZE, BOX, emptyGrid, deepCopy, boxOrigin, candidates, isValidPlacement,
    isComplete, conflictsCount, generateFullSolution, generatePuzzle, parseGrid, shuffle,
  };
})();

File: sudoku_ai/board.py
"""Core Sudoku board representation, validation, and puzzle generation.

All solvers operate on a 9x9 numpy-like list-of-lists grid where 0 means
"empty". This module is the shared substrate every AI approach builds on.
"""
import random
from copy import deepcopy

SIZE = 9
BOX = 3


def empty_grid():
    return [[0] * SIZE for _ in range(SIZE)]


def in_bounds_box(r, c):
    br, bc = (r // BOX) * BOX, (c // BOX) * BOX
    return br, bc


def candidates(grid, r, c):
    if grid[r][c] != 0:
        return set()
    used = set(grid[r]) | {grid[i][c] for i in range(SIZE)}
    br, bc = in_bounds_box(r, c)
    for i in range(br, br + BOX):
        for j in range(bc, bc + BOX):
            used.add(grid[i][j])
    return set(range(1, 10)) - used


def is_valid_placement(grid, r, c, val):
    if val == 0:
        return True
    for i in range(SIZE):
        if i != c and grid[r][i] == val:
            return False
        if i != r and grid[i][c] == val:
            return False
    br, bc = in_bounds_box(r, c)
    for i in range(br, br + BOX):
        for j in range(bc, bc + BOX):
            if (i, j) != (r, c) and grid[i][j] == val:
                return False
    return True


def is_complete(grid):
    return all(grid[r][c] != 0 for r in range(SIZE) for c in range(SIZE))


def is_solved(grid):
    if not is_complete(grid):
        return False
    for r in range(SIZE):
        for c in range(SIZE):
            v = grid[r][c]
            grid[r][c] = 0
            ok = is_valid_placement(grid, r, c, v)
            grid[r][c] = v
            if not ok:
                return False
    return True


def conflicts_count(grid):
    """Total number of rule violations. 0 means fully consistent (used as
    the fitness/cost signal shared by the genetic, annealing and RL solvers).
    """
    count = 0
    for i in range(SIZE):
        row = [v for v in grid[i] if v != 0]
        col = [grid[r][i] for r in range(SIZE) if grid[r][i] != 0]
        count += len(row) - len(set(row))
        count += len(col) - len(set(col))
    for br in range(0, SIZE, BOX):
        for bc in range(0, SIZE, BOX):
            vals = [grid[r][c] for r in range(br, br + BOX) for c in range(bc, bc + BOX) if grid[r][c] != 0]
            count += len(vals) - len(set(vals))
    return count


def _solve_backtracking(grid):
    """Classic exact solver, used only to manufacture solvable puzzles
    and to verify/ground-truth other solvers. See solvers/backtracking.py
    for the documented, user-facing version of this algorithm.
    """
    for r in range(SIZE):
        for c in range(SIZE):
            if grid[r][c] == 0:
                for val in candidates(grid, r, c):
                    grid[r][c] = val
                    if _solve_backtracking(grid):
                        return True
                    grid[r][c] = 0
                return False
    return True


def generate_full_solution():
    grid = empty_grid()
    nums = list(range(1, 10))
    for box in range(0, SIZE, BOX):
        random.shuffle(nums)
        idx = 0
        for i in range(BOX):
            for j in range(BOX):
                grid[box + i][box + j] = nums[idx]
                idx += 1
    _solve_backtracking(grid)
    return grid


def generate_puzzle(num_clues=32):
    """Generate a random solvable puzzle with roughly `num_clues` filled cells."""
    solution = generate_full_solution()
    puzzle = deepcopy(solution)
    cells = [(r, c) for r in range(SIZE) for c in range(SIZE)]
    random.shuffle(cells)
    to_remove = SIZE * SIZE - num_clues
    for r, c in cells[:to_remove]:
        puzzle[r][c] = 0
    return puzzle, solution


def grid_to_str(grid):
    return "\n".join(" ".join(str(v) for v in row) for row in grid)


def parse_grid(text):
    """Parse a grid from a string of 81 digits (0 or . for blank), row-major."""
    digits = [ch for ch in text if ch.isdigit() or ch == "."]
    if len(digits) != 81:
        raise ValueError(f"Expected 81 cells, got {len(digits)}")
    grid = empty_grid()
    for idx, ch in enumerate(digits):
        r, c = divmod(idx, SIZE)
        grid[r][c] = 0 if ch == "." else int(ch)
    return grid
